WalletWasabi.Tests/UnitTests/BranchAndBoundTests.cs

```diff
@@ -0,0 +1,106 @@
+			BranchAndBound selector = new(utxos);
+			List<SmartCoin> expectedCoins = new() { BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(50000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(40000)) };
+			Money target = Money.Satoshis(190000);
+
+			bool wasSuccessful = selector.TryGetExactMatch(target, out List<SmartCoin>? selectedCoins);
+
+			Assert.True(wasSuccessful);
+			Assert.NotNull(selectedCoins);
+			Assert.Equal(expectedCoins, selectedCoins);
+		}
+
+		[Fact]
+		public void CanSelectEveryCoin()
+		{
+			List<SmartCoin> utxos = new() { BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(120000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)) };
+			BranchAndBound selector = new(utxos);
+			List<SmartCoin> expectedCoins = new() { BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(120000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)) };
+			Money target = Money.Satoshis(320000);
+
+			bool wasSuccessful = selector.TryGetExactMatch(target, out List<SmartCoin>? selectedCoins);
+
+			Assert.True(wasSuccessful);
+			Assert.NotNull(selectedCoins);
+			Assert.Equal(expectedCoins, selectedCoins);
+		}
+
+		[Fact]
+		public void TargetIsBiggerThanBalance()
+		{
+			List<SmartCoin> utxos = GenList();
+			BranchAndBound selector = new(utxos);
+			Money target = Money.Satoshis(11111111111111111);
+
+			bool wasSuccessful = selector.TryGetExactMatch(target, out List<SmartCoin>? selectedCoins);
+
+			Assert.False(wasSuccessful);
+			Assert.Null(selectedCoins);
+		}
+
+		[Fact]
+		public void ReturnNullIfNoExactMatchFoundTest()
+		{
+			List<SmartCoin> utxos = new() { BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(120000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(100000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(50000)), BitcoinFactory.CreateSmartCoin(HdPubKey, Money.Satoshis(40000)) };
+			BranchAndBound selector = new(utxos);
+			Money target = Money.Satoshis(300000);
+
+			bool wasSuccessful = selector.TryGetExactMatch(target, out List<SmartCoin>? selectedCoins);
+
+			Assert.False(wasSuccessful);
+			Assert.Null(selectedCoins);
+		}
+
+		private List<SmartCoin> GenList()
+		{
+			KeyManager keyManager = ServiceFactory.CreateKeyManager();
+			HdPubKey hdPubKey = BitcoinFactory.CreateHdPubKey(keyManager);
+			List<SmartCoin> availableCoins = new();
+
+			for (int i = 0; i < 1000; i++)
+			{
+				availableCoins.Add(BitcoinFactory.CreateSmartCoin(hdPubKey, (ulong)Random.Next((int)Money.Satoshis(1000), (int)Money.Satoshis(99999999))));
```

Comment:
`(int)Money.Satoshis(1000)` is a fancy way (and ineffective way) of writing `1000` imho, so you can just write:

```suggestion
				ulong sats = (ulong)Random.Next(1000, 99999999);
				availableCoins.Add(BitcoinFactory.CreateSmartCoin(hdPubKey, amount: sats));
```
-or-

```suggestion
				availableCoins.Add(BitcoinFactory.CreateSmartCoin(hdPubKey, amount: (ulong)Random.Next(1000, 99999999)));
```

Reply:
You are right, that's more complicated than it should be.